### ISSUE TITLE
Use pretty printing library such as pprint

### ISSUE THREAD

**Issue by julien-truffaut** [tpolecat/tut#83]
It is sometimes difficult to read the output of the REPL, could it be possible to use something like [pprint](http://lihaoyi.github.io/upickle-pprint/pprint/)

**Comment by tpolecat** [comment]
The issue here (as discussed on gitter, repeated here for the record) is that **tut** just hooks into an `IMain` instance which just spits stuff to an output stream; we never see the results of an evaluation. So to do this we would need to use Ammonite or the script engine or something. Worth keeping in mind if we do a rewrite.

**Comment by davegurnell** [comment]
Can it be reimplemented on top of Ammonite? That seems to pretty print values pretty well. I don't think it pretty prints types yet, but it's a step in the right direction.

**Comment by tpolecat** [comment]
I tried Ammonite a couple times and it blew up ([here](https://github.com/tpolecat/doobie/issues/258) and [here](https://github.com/lihaoyi/Ammonite/issues/280)) so I don't think it's quite ready to deal with the kinds of types that I need to support for my personal work. But this may just be a matter of contributing some fixes; I haven't looked at the implementation at all.

I also am somewhat hesitant to change the look and feel because the intent is that **tut** should present realistic REPL interactions. Maybe there is some kind of middle ground.

This issue is related to #33.

**Comment by davegurnell** [comment]
Understood. As you know, @d6y, @noelwelsh and I are interested in using tut for our books at Underscore, so we're a bit of an odd customer. We'd be happy if we could write a thin wrapper around core tut, to customise the output without messing up the out-of-the-box behaviour.

**Comment by ryan-williams** [comment]
I'm poking at letting Tut (optionally) use an ammonite repl instead of IMain; the two ammonite issues cited [above](https://github.com/tpolecat/tut/issues/83#issuecomment-167813376) seem long fixed.

Tut's interactions with IMain also seem [quite](https://github.com/tpolecat/tut/blob/v0.6.3/modules/core/src/main/scala/tut/Tut.scala#L31-L32) [minimal](https://github.com/tpolecat/tut/blob/v0.6.3/modules/core/src/main/scala/tut/Tut.scala#L52), so I'm hopeful this is possible / not too hard.

**Comment by ryan-williams** [comment]
Turns out I was mistaken in thinking that Ammonite/PPrint would support customizable pretty-printing for various types (https://github.com/lihaoyi/PPrint/issues/11), which somewhat saps my motivation to work on this 🙁

**Comment by tpolecat** [comment]
At the risk of sounding like a crazy person, making the compiler itself pluggable in this regard seems like a better strategy. We just need to nerdsnipe someone into figuring out how to do it.

**Comment by ryan-williams** [comment]
That sounds good… but at the risk of sounding like a crazier person, I am going to work in parallel on a version of tut that e.g. wraps lines in `println(_.show)`, using customizable `cats.Show` instances, possibly based on ammonite since it seems to offer better control over some of its output/error/result streams…

**Comment by tpolecat** [comment]
👍 You can do better with something like `trait TPrint[T] { val toString: String }` and then `tshow[T: TPrint: Show](t: T): String = ...` which is kind of what Ammonite used to do.

**Comment by ryan-williams** [comment]
Making sure I'm following: you've formalized customizable display of both the type and value, piggy-backing on PPrint's existing `TPrint` type-class for the former, and using `cats.Show` for the latter?

**Comment by tpolecat** [comment]
Yeah just saying you can derive pretty-printers for types too. Whether with Ammonite/PPrint or without.

**Comment by ryan-williams** [comment]
Taking a break from this for today; [here's a branch](https://github.com/ryan-williams/tut/tree/amm) with some minimal progress:

- Ammonite [`Repl`](https://github.com/ryan-williams/tut/commit/4d7c44d9eb1ff58067cbb8148c71f06f8a6c548b#diff-696aedea3a9d6cc94d6c76c08900705f) copy-pasted from [`TestRepl`](https://github.com/lihaoyi/Ammonite/blob/1.0.5/amm/repl/src/test/scala/ammonite/TestRepl.scala)
- [a few tests](https://github.com/ryan-williams/tut/commit/4d7c44d9eb1ff58067cbb8148c71f06f8a6c548b#diff-edaf2444d579a320b87d3ebc9ed54dd3) showing ways to display results using custom `cats.Show` instances
- [half-baked interface](https://github.com/ryan-williams/tut/commit/4d7c44d9eb1ff58067cbb8148c71f06f8a6c548b#diff-ef364c8595e43cf28fc4026a299d75e7) over `IMain` or `Repl`
  - plumbing printstreams through usage sites not done
